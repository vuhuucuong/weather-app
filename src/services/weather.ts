import axios from 'axios'
import { GetWeatherResponse, SearchLocationResponse } from '../types/weather'

interface SearchLocationParams {
  queryType: 'text' | 'lattlong'
  queryString?: string
}

export const searchLocation = async ({
  queryType,
  queryString,
}: SearchLocationParams): Promise<SearchLocationResponse> => {
  const queryKey = queryType === 'text' ? 'query' : 'lattlong'
  const { data } = await axios.get(
    `http://localhost:3000/api/location/search/?${queryKey}=${queryString}`
  )
  return data
}

export const getWeather = async ({
  woeid,
}: {
  woeid: number
}): Promise<GetWeatherResponse> => {
  const { data } = await axios.get(
    `http://localhost:3000/api/location/${woeid}`
  )
  return data
}
