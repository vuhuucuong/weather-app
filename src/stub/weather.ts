import {
  SearchLocationResponse,
  GetWeatherResponse,
  WeatherInfo,
} from '../types/weather'

export const searchResponse: SearchLocationResponse = [
  {
    title: 'Ho Chi Minh City',
    location_type: 'City',
    woeid: 1252431,
    latt_long: '10.759180,106.662498',
  },
  {
    title: 'Houston',
    location_type: 'City',
    woeid: 2424766,
    latt_long: '29.760450,-95.369781',
  },
  {
    title: 'Hong Kong',
    location_type: 'City',
    woeid: 2165352,
    latt_long: '22.411200,114.153999',
  },
  {
    title: 'Phoenix',
    location_type: 'City',
    woeid: 2471390,
    latt_long: '33.448261,-112.075768',
  },
  {
    title: 'Guangzhou',
    location_type: 'City',
    woeid: 2161838,
    latt_long: '23.107389,113.267616',
  },
  {
    title: 'Oklahoma City',
    location_type: 'City',
    woeid: 2464592,
    latt_long: '35.472000,-97.520348',
  },
  {
    title: 'Honolulu',
    location_type: 'City',
    woeid: 2423945,
    latt_long: '21.304850,-157.857758',
  },
  {
    title: 'Anchorage',
    location_type: 'City',
    woeid: 2354490,
    latt_long: '61.217548,-149.858383',
  },
  {
    title: 'Stockholm',
    location_type: 'City',
    woeid: 906057,
    latt_long: '59.332169,18.062429',
  },
  {
    title: 'Lahore',
    location_type: 'City',
    woeid: 2211177,
    latt_long: '31.549910,74.327301',
  },
  {
    title: 'Lake Tahoe',
    location_type: 'City',
    woeid: 23511744,
    latt_long: '39.021400,-120.044823',
  },
  {
    title: 'Hangzhou',
    location_type: 'City',
    woeid: 2132574,
    latt_long: '30.252501,120.165024',
  },
  {
    title: 'Windhoek',
    location_type: 'City',
    woeid: 1466719,
    latt_long: '-22.559561,17.08445',
  },
]

export const weatherListResponse: GetWeatherResponse = {
  consolidated_weather: [
    {
      id: 6467278154498048,
      weather_state_name: 'Clear',
      weather_state_abbr: 'c',
      wind_direction_compass: 'W',
      created: '2021-01-12T16:00:51.209616Z',
      applicable_date: '2021-01-13',
      min_temp: 19.045,
      max_temp: 30.015,
      the_temp: 28.395000000000003,
      wind_speed: 3.579970911861396,
      wind_direction: 266.85470957018885,
      air_pressure: 1013.0,
      humidity: 55,
      visibility: 14.852946293645113,
      predictability: 68,
    },
    {
      id: 5101955580952576,
      weather_state_name: 'Light Cloud',
      weather_state_abbr: 'lc',
      wind_direction_compass: 'SW',
      created: '2021-01-12T16:00:54.113949Z',
      applicable_date: '2021-01-14',
      min_temp: 20.020000000000003,
      max_temp: 31.11,
      the_temp: 29.814999999999998,
      wind_speed: 2.6061191839497337,
      wind_direction: 213.91194466624782,
      air_pressure: 1012.0,
      humidity: 60,
      visibility: 14.79267328799809,
      predictability: 70,
    },
    {
      id: 5985912178606080,
      weather_state_name: 'Light Cloud',
      weather_state_abbr: 'lc',
      wind_direction_compass: 'WNW',
      created: '2021-01-12T16:00:57.251156Z',
      applicable_date: '2021-01-15',
      min_temp: 21.285,
      max_temp: 31.964999999999996,
      the_temp: 30.939999999999998,
      wind_speed: 3.517598026864445,
      wind_direction: 295.8320442104878,
      air_pressure: 1010.0,
      humidity: 58,
      visibility: 14.929374950290304,
      predictability: 70,
    },
    {
      id: 5069710241562624,
      weather_state_name: 'Heavy Cloud',
      weather_state_abbr: 'hc',
      wind_direction_compass: 'NNE',
      created: '2021-01-12T16:01:00.582943Z',
      applicable_date: '2021-01-16',
      min_temp: 22.22,
      max_temp: 31.994999999999997,
      the_temp: 31.5,
      wind_speed: 3.7071944303454494,
      wind_direction: 11.475268061247151,
      air_pressure: 1009.0,
      humidity: 59,
      visibility: 15.243478087966277,
      predictability: 71,
    },
    {
      id: 5096164488642560,
      weather_state_name: 'Light Rain',
      weather_state_abbr: 'lr',
      wind_direction_compass: 'NE',
      created: '2021-01-12T16:01:03.196090Z',
      applicable_date: '2021-01-17',
      min_temp: 22.63,
      max_temp: 29.505000000000003,
      the_temp: 27.6,
      wind_speed: 2.8615626739839337,
      wind_direction: 45.5,
      air_pressure: 1011.0,
      humidity: 62,
      visibility: 9.999726596675416,
      predictability: 75,
    },
    {
      id: 4531189575581696,
      weather_state_name: 'Light Cloud',
      weather_state_abbr: 'lc',
      wind_direction_compass: 'N',
      created: '2021-01-12T16:01:06.183334Z',
      applicable_date: '2021-01-18',
      min_temp: 22.33,
      max_temp: 32.12,
      the_temp: 32.66,
      wind_speed: 4.000494114372067,
      wind_direction: 4.5,
      air_pressure: 1010.0,
      humidity: 50,
      visibility: 9.999726596675416,
      predictability: 70,
    },
  ],
  time: '2021-01-13T00:58:36.569288+07:00',
  sun_rise: '2021-01-13T06:15:38.919814+07:00',
  sun_set: '2021-01-13T17:48:40.645555+07:00',
  timezone_name: 'LMT',
  parent: {
    title: 'Vietnam',
    location_type: 'Country',
    woeid: 23424984,
    latt_long: '15.974210,107.868042',
  },
  sources: [
    {
      title: 'BBC',
      slug: 'bbc',
      url: 'http://www.bbc.co.uk/weather/',
      crawl_rate: 360,
    },
    {
      title: 'Forecast.io',
      slug: 'forecast-io',
      url: 'http://forecast.io/',
      crawl_rate: 480,
    },
    {
      title: 'Met Office',
      slug: 'met-office',
      url: 'http://www.metoffice.gov.uk/',
      crawl_rate: 180,
    },
    {
      title: 'OpenWeatherMap',
      slug: 'openweathermap',
      url: 'http://openweathermap.org/',
      crawl_rate: 360,
    },
    {
      title: 'World Weather Online',
      slug: 'world-weather-online',
      url: 'http://www.worldweatheronline.com/',
      crawl_rate: 360,
    },
  ],
  title: 'Ho Chi Minh City',
  location_type: 'City',
  woeid: 1252431,
  latt_long: '10.759180,106.662498',
  timezone: 'Asia/Ho_Chi_Minh',
}

export const transformedWeatherList: WeatherInfo[] = [
  {
    title: 'Ho Chi Minh City',
    weatherStateName: 'Clear',
    theTemp: 28.395000000000003,
    minTemp: 19.045,
    maxTemp: 30.015,
    applicableDate: '2021-01-13',
    weatherStateAbbr: 'c',
  },
  {
    title: 'Ho Chi Minh City',
    weatherStateName: 'Light Cloud',
    theTemp: 29.814999999999998,
    minTemp: 20.020000000000003,
    maxTemp: 31.11,
    applicableDate: '2021-01-14',
    weatherStateAbbr: 'lc',
  },
  {
    title: 'Ho Chi Minh City',
    weatherStateName: 'Light Cloud',
    theTemp: 30.939999999999998,
    minTemp: 21.285,
    maxTemp: 31.964999999999996,
    applicableDate: '2021-01-15',
    weatherStateAbbr: 'lc',
  },
  {
    title: 'Ho Chi Minh City',
    weatherStateName: 'Heavy Cloud',
    theTemp: 31.5,
    minTemp: 22.22,
    maxTemp: 31.994999999999997,
    applicableDate: '2021-01-16',
    weatherStateAbbr: 'hc',
  },
  {
    title: 'Ho Chi Minh City',
    weatherStateName: 'Light Rain',
    theTemp: 27.6,
    minTemp: 22.63,
    maxTemp: 29.505000000000003,
    applicableDate: '2021-01-17',
    weatherStateAbbr: 'lr',
  },
  {
    title: 'Ho Chi Minh City',
    weatherStateName: 'Light Cloud',
    theTemp: 32.66,
    minTemp: 22.33,
    maxTemp: 32.12,
    applicableDate: '2021-01-18',
    weatherStateAbbr: 'lc',
  },
]
