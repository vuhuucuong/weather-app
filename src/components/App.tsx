import React, { useEffect, useState } from 'react'
import styles from './App.module.css'
import { WeatherInfo } from '../types/weather'
import SearchBar from './SearchBar'
import WeatherCard from './WeatherCard'
import CurrentDayBlock from './CurrentDayWeather'
import WeekDayWeatherList from './WeekDayWeatherList'
import { getCurrentDayWeather } from '../utils'
import Spinner from './Spinner'

function App() {
  const [selectedWeather, setSelectedWeather] = useState<WeatherInfo | null>(
    null
  )
  const [listWeathers, setListWeathers] = useState<WeatherInfo[]>([])
  const [isLoading, setIsLoading] = useState<boolean>(false)

  useEffect(() => {
    if (listWeathers.length > 0) {
      const currentDayWeather = getCurrentDayWeather(listWeathers)
      setSelectedWeather(currentDayWeather)
    }
  }, [listWeathers])

  const renderCardContent = () => {
    if (isLoading) {
      return (
        <div className="text-center">
          <Spinner size={50} />
        </div>
      )
    }
    if (!selectedWeather) {
      return <h4 className="text-center">Please enter a city name to search</h4>
    }
    return (
      <>
        <CurrentDayBlock weatherInfoCurrentDay={selectedWeather} />
        <hr />
        <WeekDayWeatherList
          selectedWeather={selectedWeather}
          setSelectedWeather={setSelectedWeather}
          weatherInfoList={listWeathers}
        />
      </>
    )
  }

  return (
    <div className={styles.appWrapper}>
      <div className={styles.innerContainer}>
        <SearchBar
          setListWeathers={setListWeathers}
          setIsLoading={setIsLoading}
        />
        <WeatherCard>{renderCardContent()}</WeatherCard>
      </div>
    </div>
  )
}

export default App
