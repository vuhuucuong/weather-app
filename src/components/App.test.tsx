import { render, screen, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { searchResponse, weatherListResponse } from '../stub/weather'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import App from './App'

const server = setupServer(
  rest.get('/api/location/search/*', (req, res, ctx) => {
    return res(ctx.json(searchResponse))
  }),
  rest.get('/api/location/*', (req, res, ctx) => {
    return res(ctx.json(weatherListResponse))
  })
)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

it('should match snapshot', () => {
  const tree = renderer.create(<App />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders without crashing', () => {
  render(<App />)
  expect(
    screen.getByText('Please enter a city name to search')
  ).toBeInTheDocument()
})

it('can search for a location and show data', async () => {
  render(<App />)
  fireEvent.change(screen.getByTestId('searchInput'), {
    target: { value: 'test' },
  })
  expect((screen.getByTestId('searchInput') as HTMLInputElement).value).toBe(
    'test'
  )
  await screen.findAllByTestId('dropDownItem')
  expect(screen.getAllByTestId('dropDownItem')).toHaveLength(
    searchResponse.length
  )
  fireEvent.click(screen.getByText(searchResponse[0].title))
  await screen.findByTestId('currentDayWeather')
})
