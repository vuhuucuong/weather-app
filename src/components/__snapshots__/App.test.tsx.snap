// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="appWrapper"
>
  <div
    className="innerContainer"
  >
    <div
      className="w-100 position-relative"
    >
      <div
        className="input-group"
      >
        <input
          className="form-control"
          data-testid="searchInput"
          id="searchInput"
          list="cityDataList"
          onChange={[Function]}
          placeholder="Enter city name..."
          value=""
        />
      </div>
      <ul
        className="list-group w-100 position-absolute overflow-auto dropDown"
        data-element-type="dropDown"
        data-testid="dropDown"
        onClick={[Function]}
      />
    </div>
    <div
      className="d-flex flex-column justify-content-center text-light p-5 border-0 shadow-sm dayWeatherCard"
    >
      <h4
        className="text-center"
      >
        Please enter a city name to search
      </h4>
    </div>
  </div>
</div>
`;
